rfcs/20200804-configurable-filesystems.md

```diff
@@ -0,0 +1,135 @@
+# Configurable File Systems
+
+| Status        | Proposed                                                                                      |
+| :------------ | :-------------------------------------------------------------------------------------------- |
+| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) (update when you have community PR #) |
+| **Author(s)** | Sami Kama (kamsami@amazon.com)                                                                |
+| **Sponsor**   | Mihai Maruseac (mihaimaruseac@google.com)                                                     |
+| **Updated**   | 2020-08-04                                                                                    |
+
+## Objective
+
+The aim of this RFC to extend filesystem API to enable users to pass configuration parameters to tune the behavior of implementation to their use cases.
+
+## Motivation
+
+There are many FileSystem implementations in Tensorflow that enable interaction with various storage solutions. Most of these implementations have internal parameters that are suitable for generic use case but not necessarily optimal for all cases. For example accessing remote filesystems through multiple threads can improve the throughput if there is a high bandwidth connection to the remote thus increasing number of connections might be beneficial. On the other hand if the connection is slow, a higher number of threads will just waste resources and may even reduce the throughput. Depending on the resources available during the execution, users should be able to alter some of the parameters of the Filesystems to improve the performance of their execution. This can be especially useful for the the cases where the execution is data i/o bound.
```

Comment:
What's the relation between this and SIG IO's filesystem extensions? @mihaimaruseac @samikama @tensorflow/sig-io-maintainers

Reply:
@terrytangyuan Are you talking about plugin based filesystem or something else? Could you please be more specific?

Reply:
I mean the existing file system extensions in [TF IO kernels](https://github.com/tensorflow/io/tree/4b02ef9113de55450f49261a8bfee108c5c5fa74/tensorflow_io/core/kernels), e.g. gstpu, oss, azure, etc. For example, we can use Azure like the following ([full tutorial](https://www.tensorflow.org/io/tutorials/azure)):
```
pathname = 'az://{}/aztest'.format(account_name)
tf.io.gfile.mkdir(pathname)
```

Reply:
Hi @terrytangyuan,

Sorry if the document was not clear. This proposal do not change existing behavior. It extends existing api so that you can do things like @martinwicke mentioned above.

Reply:
Got it. Perhaps mentioning the relationship with TF IO would avoid some confusion. Thanks.